core/include/vecmem/containers/impl/static_array.ipp

```diff
@@ -122,20 +122,141 @@ static_array<T, N>::data(void) const {
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::const_iterator
+static_array<T, N>::cbegin() const {
+
+    return begin();
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::iterator
+static_array<T, N>::end() {
+
+    return m_array + N;
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::const_iterator
+static_array<T, N>::end() const {
+
+    return m_array + N;
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::const_iterator
+static_array<T, N>::cend() const {
+
+    return end();
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::reverse_iterator
+static_array<T, N>::rbegin() {
+
+    return reverse_iterator(end());
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr
+    typename static_array<T, N>::const_reverse_iterator
+    static_array<T, N>::rbegin() const {
+
+    return const_reverse_iterator(end());
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr
+    typename static_array<T, N>::const_reverse_iterator
+    static_array<T, N>::crbegin() const {
+
+    return rbegin();
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::reverse_iterator
+static_array<T, N>::rend() {
+
+    return reverse_iterator(begin());
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr
+    typename static_array<T, N>::const_reverse_iterator
+    static_array<T, N>::rend() const {
+
+    return const_reverse_iterator(begin());
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr
+    typename static_array<T, N>::const_reverse_iterator
+    static_array<T, N>::crend() const {
+
+    return rend();
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr bool static_array<T, N>::empty() const {
+
+    return N == 0;
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::size_type
+static_array<T, N>::size() const {
+
+    return N;
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::size_type
+static_array<T, N>::max_size() const {
+
+    return N;
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE void static_array<T, N>::fill(const_reference value) {
+
+    for (std::size_t i = 0; i < N; ++i) {
+        m_array[i] = value;
+    }
+}
+
+template <typename T, std::size_t N>
+template <typename Tp1, typename... Tp>
+VECMEM_HOST_AND_DEVICE void static_array<T, N>::static_array_impl(size_type i,
+                                                                  Tp1&& a1,
+                                                                  Tp&&... a) {
+
+    m_array[i] = a1;
+    static_array_impl(i + 1, std::forward<Tp>(a)...);
```

Comment:
Honestly I don't think this is too big of a problem. This method is most likely used only for relatively small arrays, so that in combination with tail-call elimination will hopefully make this perform quite well. Hopefully.